src/com/maddyhome/idea/vim/group/CopyGroup.java

```diff
@@ -104,141 +105,249 @@ public boolean yankRange(@NotNull Editor editor, @Nullable TextRange range, @Not
       }
-      else {
-        pos = editor.getCaretModel().getOffset();
-        if (!EditorHelper.isLineEmpty(editor, editor.getCaretModel().getLogicalPosition().line, false)) {
-          pos++;
-        }
+      else if (subMode != CommandState.SubMode.VISUAL_LINE) {
+        editor.getDocument().insertString(startOffset, "\n");
+        ++startOffset;
       }
-      // In case when text is empty this can occur
-      if (pos > 0 && pos > editor.getDocument().getTextLength()) {
-        pos--;
+    }
+    else if (selectionType != SelectionType.CHARACTER_WISE) {
+      if (subMode == CommandState.SubMode.VISUAL_LINE) {
+        editor.getDocument().insertString(startOffset, "\n");
       }
-      putText(editor, context, pos, StringUtil.notNullize(reg.getText()), reg.getType(), count, indent, cursorAfter,
-              CommandState.SubMode.NONE);
-
-      return true;
     }
 
-    return false;
+    final String text = indent ?
+        StringUtil.notNullize(register.getText()) :
+        removeLeadingSpaces(StringUtil.notNullize(register.getText()));
+    final int endOffset = selectionType == SelectionType.BLOCK_WISE ?
+        putTextBlockwise(editor, caret, text, count, startOffset) :
+        putText(editor, caret, text, count, startOffset);
+
+    VimPlugin.getRegister().storeText(editor, new TextRange(startOffset, startOffset + (endOffset - startOffset) / count),
+        selectionType, false);
+
+    moveCursorToOffset(editor, caret, selectionType, startOffset, endOffset, cursorAfter);
+
+    return true;
   }
 
-  public boolean putVisualRange(@NotNull Editor editor, @NotNull DataContext context, @NotNull TextRange range,
-                                int count, boolean indent, boolean cursorAfter) {
-    CommandState.SubMode subMode = CommandState.getInstance(editor).getSubMode();
-    Register reg = VimPlugin.getRegister().getLastRegister();
-    // Without this reset, the deleted text goes into the same register we just pasted from.
-    VimPlugin.getRegister().resetRegister();
-    if (reg != null) {
-      final SelectionType type = reg.getType();
-      if (type == SelectionType.LINE_WISE && editor.isOneLineMode()) {
-        return false;
+  @NotNull
+  private String removeLeadingSpaces(@NotNull String s) {
+    final StringBuilder ret = new StringBuilder(s.length());
+    int idx = 0;
+    while (idx < s.length() && Character.isWhitespace(s.charAt(idx))) {
+      if (s.charAt(idx) != ' ') {
+        ret.append(s.charAt(idx));
       }
+      ++idx;
+    }
+    return ret.append(s.substring(idx)).toString();
+  }
 
-      int start = range.getStartOffset();
-      int end = range.getEndOffset();
-      int endLine = editor.offsetToLogicalPosition(end).line;
-      if (logger.isDebugEnabled()) {
-        logger.debug("start=" + start);
-        logger.debug("end=" + end);
+
+  private int putText(@NotNull Editor editor, @NotNull Caret caret, @NotNull String text, int count, int startOffset) {
```

Comment:
The refactoring of the original `putText()` looks good if we look at its decomposition into separate methods. Unfortunately, some of the original logic inside `putText()` was removed. Let's add tests for all the logic that was removed, or, even better, do this refactoring once again: initially without any changes to the logic, then you could proceed with removing some pieced of `putText()` logic, but only with tests for these pieces.